core/git_command.py

```diff
@@ -245,38 +254,40 @@ def _repo_path(self, throw_on_stderr=True):
         Return the absolute path to the git repo that contains the file that this
         view interacts with.  Like `file_path`, this can be overridden by setting
         the view's `git_savvy.repo_path` setting.
-        """
-        def invalid_repo():
-            if throw_on_stderr:
-                raise ValueError("Unable to determine Git repo path.")
-            return None
 
+        Do not raise error when throw_on_stderr is `False`. It is needed
+        in GsUpdateStatusBarCommand, otherwise, spurious popup will be shown.
+        """
         # The below condition will be true if run from a WindowCommand and false
         # from a TextCommand.
         view = self.window.active_view() if hasattr(self, "window") else self.view
+        window = view.window()
         repo_path = view.settings().get("git_savvy.repo_path")
 
         if not repo_path or not os.path.exists(repo_path):
-            working_dir = self.find_working_dir()
-            if not working_dir or not os.path.isdir(working_dir):
-                return invalid_repo()
-
-            stdout = self.git(
-                "rev-parse",
-                "--show-toplevel",
-                working_dir=working_dir,
-                throw_on_stderr=throw_on_stderr
-                )
+            file_name = view.file_name()
 
-            repo_path = stdout.strip()
+            # try the current file first
+            if file_name:
+                repo_path = self.find_git_toplevel(
+                    os.path.dirname(file_name), throw_on_stderr=False)
+                # only set `git_savvy.repo_path` for current file
+                if repo_path:
+                    view.settings().set("git_savvy.repo_path", os.path.realpath(repo_path))
 
+            # fallback: use the first folder if the current file is not inside a git repo
             if not repo_path:
-                return invalid_repo()
+                if window:
+                    folders = window.folders()
+                    if folders:
+                        repo_path = self.find_git_toplevel(
+                            folders[0], throw_on_stderr=throw_on_stderr)
```

Comment:
Another use of `find_working_dir` is to find a [directory][1] to initialize. In there, we do not assume that that folder is a git repo. We might end up putting a bunch of different logic into `find_working_dir`, something like `find_working_dir(find_git_root=True)` which I find is of the same complexity to separate them into two functions.


>  if the .bashrc is in a git repo (e.g a dotfiles repo), then the git commands should run on this repo

if it is in a git repo, the [first if clause][2] had already return the repo. The logic is triggered only for files out of git repo. Setting `git_savvy.repo_path` is a shortcut for files staying in a git repo.




[1]: https://github.com/divmain/GitSavvy/blob/41f1d7b2d6ceaecff3c896ed82891aff94afa4a6/core/commands/init.py#L64-L64

[2]: https://github.com/divmain/GitSavvy/blob/41f1d7b2d6ceaecff3c896ed82891aff94afa4a6/core/git_command.py#L271-L271